File: DynamicProgramming/minimal palindromic decomposition/minimal_palindromic_decomposition.c
/**********************************************************
 *
 *  Find Minimal Palindromic Decomposition
 *
 *********************************************************/
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MIN(i, j)   ((i) < (j) ? (i) : (j))

int findMinimalPalindromicDecomposition(const char *string) {
    int length = strlen(string);

    // isPalindrom[i][j] == 1 iff substring(string, i, j) is a palindrome
    char *isPalindrom = malloc(length * length);
    for(int i = length - 1; i >= 0; --i) {
        for(int j = i; j < length; ++j) {
            isPalindrom[i * length + j] = 0;

            if (string[i] != string[j])
                continue;

            if ((j - i <= 1) || isPalindrom[(i + 1) * length + (j - 1)])
                isPalindrom[i * length + j] = 1;
        }
    }   

    int *minimumPalindromicDecomposition = malloc(sizeof(int) * length);
    for(int i = length - 1; i >= 0; --i) {
        int min = length;
        for(int j = i; j < length; ++j) {
            if (isPalindrom[i * length + j])
                min = j + 1 == length ? 1 : MIN(min, 1 + minimumPalindromicDecomposition[j + 1]);
        }
        minimumPalindromicDecomposition[i] = min;
    }
    
    free(isPalindrom);
    int result = minimumPalindromicDecomposition[0]; 
    free(minimumPalindromicDecomposition);
    return result;
}

typedef struct {
    char *string;
    int result;
} TestCase;

TestCase tests[] = {
    {"1234", 4},        // 1|2|3|4
    {"11234", 4},       // 11|2|3|4
    {"1211213121", 2}   //121|1213121
};
  
#define NUMBER_OF_TESTS sizeof(tests) / sizeof(TestCase)

int main() {
    int errors = 0;
    for(int i = 0; i < NUMBER_OF_TESTS; ++i) {
        int result = findMinimalPalindromicDecomposition(tests[i].string);
        if (result != tests[i].result) {
            printf("Error: test case number %d returned %d. Expected %d\n",
                i + 1, result, tests[i].result);
            errors++;
        } 
    }

    if (errors)
        printf("Got %d errors\n", errors);
    else
        printf("Good work\n");
    return 0;
} 

File: DynamicProgramming/minimal palindromic decomposition/MinimalPalindromicDecomposition.java
/**********************************************************
 * Find Minimal Palindromic Decomposition
 *********************************************************/

public class MinimalPalindromicDecomposition {
    public static int recursion(String s, int index) {
        int length = s.length();
        boolean isPalindrome[][] = calcPalindromes(s);

        if (index == length)
            return 0;

        int min = length;
        for (int j = index; j < length; ++j)
            if (isPalindrome[index][j])
                min = Math.min(min, 1 + recursion(s, j + 1));

        return min;
    }

    public static int dynamicProgramming(String string) {
        int length = string.length();
        boolean isPalindrome[][] = calcPalindromes(string);

        int mPD[] = new int[length + 1]; // Minimum Palindromic Decomposition

        mPD[length] = 0;

        for (int i = length - 1; i >= 0; --i) {
            int min = length;
            for (int j = i; j < length; ++j)
                if (isPalindrome[i][j])
                    min = Math.min(min, 1 + mPD[j + 1]);
            mPD[i] = min;
        }

        return mPD[0];
    }

    private static boolean[][] calcPalindromes(String s) {
        int length = s.length();
        // isPalindrome[i][j] == true iff substring(string, i, j) is a palindrome
        boolean isPalindrome[][] = new boolean[length][length];
        for (int i = length - 1; i >= 0; --i) {
            for (int j = i; j < length; ++j) {
                isPalindrome[i][j] = false;

                if (s.charAt(i) != s.charAt(j))
                    continue;

                if ((j - i <= 1) || isPalindrome[i + 1][j - 1])
                    isPalindrome[i][j] = true;
            }
        }
        return isPalindrome;
    }

    private static class TestCase {
        public TestCase(String s, int r) {
            string = s;
            result = r;
        }

        String string;
        int result;
    }

    private static TestCase tests[] = {
            new TestCase("1234", 4),        // 1|2|3|4
            new TestCase("11234", 4),       // 11|2|3|4
            new TestCase("1211213121", 2)   //121|1213121
    };

    public static void main(String[] args) {
        int errors = 0;
        for (int i = 0; i < tests.length; ++i) {
            int result = dynamicProgramming(tests[i].string);
            if (result != tests[i].result) {
                System.out.println("Error: test case number " + (i + 1) + " returned " + result + ". Expected " +
                        tests[i].result);
                errors++;
            }
        }

        if (errors > 0)
            System.out.println("Got " + errors + " errors");
        else
            System.out.println("Good work");

        errors = 0;
        for (int i = 0; i < tests.length; ++i) {
            int result = recursion(tests[i].string, 0);
            if (result != tests[i].result) {
                System.out.println("Error: test case number " + (i + 1) + " returned " + result + ". Expected " +
                        tests[i].result);
                errors++;
            }
        }

        if (errors > 0)
            System.out.println("Got " + errors + " errors");
        else
            System.out.println("Good work");
    }
}
